internal/directives/file_deleter.go

```diff
@@ -0,0 +1,139 @@
+package directives
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func init() {
+	builtins.RegisterPromotionStepRunner(newFileDeleter(), nil)
+}
+
+type fileDeleter struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+func newFileDeleter() PromotionStepRunner {
+	r := &fileDeleter{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+func (f *fileDeleter) Name() string {
+	return "delete"
+}
+
+func (f *fileDeleter) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	// Validate the configuration against the JSON Schema.
+	if err := validate(f.schemaLoader, gojsonschema.NewGoLoader(stepCtx.Config), f.Name()); err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}, err
+	}
+
+	// Convert the configuration into a typed object.
+	cfg, err := ConfigToStruct[DeleteConfig](stepCtx.Config)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not convert config into %s config: %w", f.Name(), err)
+	}
+
+	return f.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+func (f *fileDeleter) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg DeleteConfig,
+) (PromotionStepResult, error) {
+	// join with a regular join
+	absPath, err := checkAbsPath(stepCtx.WorkDir, cfg.Path)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not secure join path %q: %w", cfg.Path, err)
+	}
+
+	if checkSymlink(absPath) {
+		err := os.Remove(absPath)
+		if err != nil {
+			return PromotionStepResult{}, err
+		}
+	} else {
+		pathToDelete, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.Path)
+		if err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("could not secure join path %q: %w", cfg.Path, err)
+		}
+
+		if err = removePath(pathToDelete); err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("failed to delete %q: %w", cfg.Path, sanitizePathError(err, stepCtx.WorkDir))
+		}
+	}
+
+	return PromotionStepResult{Status: kargoapi.PromotionPhaseSucceeded}, nil
+}
+
+func checkSymlink(path string) bool {
+	fi, err := os.Lstat(path)
+	if err != nil {
+		return false
+	}
```

Comment:
This should consider that the file can already be deleted, in which case I would not expect an error to be thrown nor a removal to be attempted.

Reply:
Also if the error _isn't_ the the file/dir/symlink isn't found, it's not right to swallow it and just say "not a symlink."

This method should return such an error.

Reply:
the error returned will be of type `PathError` according to the documentation. Can you provide some more context on why should we specifically care about this error?

Reply:
> why should we specifically care about this error

It's more like, "Why _shouldn't_ we?"

As a general rule, you handle what errors you can -- in this case that entails disregarding errors that are _specifically_ that no file/dir exsists at that path -- and you surface any other problem you weren't anticipating or don't know how to handle.

Reply:
Got it, please see if this is the correct way to handle this error now.

Reply:
marking this as resolved